File: article-api/src/main/kotlin/com/sns/article/reaction/application/ReactionService.java
package com.sns.article.reaction.application;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.sns.article.reaction.domain.Reaction;
import com.sns.article.reaction.domain.ReactionNotFoundException;
import com.sns.article.reaction.domain.ReactionRepository;
import com.sns.article.reaction.domain.ReactionTarget;
import com.sns.article.reaction.domain.ReactionValidator;
import com.sns.article.reaction.dto.ReactionDto;

@Service
@Transactional(readOnly = true)
public class ReactionService {
	private final ReactionRepository reactionRepository;
	private final ReactionValidator reactionValidator;
	private final ApplicationEventPublisher applicationEventPublisher;

	public ReactionService(
		ReactionRepository reactionRepository,
		ReactionValidator reactionValidator,
		ApplicationEventPublisher applicationEventPublisher
	) {
		this.reactionRepository = reactionRepository;
		this.reactionValidator = reactionValidator;
		this.applicationEventPublisher = applicationEventPublisher;
	}

	@Transactional
	public ReactionDto create(ReactionDto dto) {
		reactionValidator.validate(dto);
		ReactionTarget target = new ReactionTarget(dto.getTargetType(), dto.getTargetId());
		Reaction reaction = reactionRepository.save(new Reaction(dto.getId(), target, dto.getType(), dto.getUserId()));
		reaction.created(applicationEventPublisher);
		reactionRepository.save(reaction);
		return ReactionDto.from(reaction);
	}

	public ReactionDto findById(Long id) {
		Reaction reaction = reactionRepository.findById(id).orElseThrow(ReactionNotFoundException::new);
		return ReactionDto.from(reaction);
	}

	@Transactional
	public void delete(Long id) {
		reactionRepository.deleteById(id);
	}
}

File: article-api/src/main/kotlin/com/sns/article/reaction/application/ReactionService.kt
package com.sns.article.reaction.application

import com.sns.article.reaction.domain.*
import com.sns.article.reaction.dto.ReactionDto
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class ReactionService(
    private val reactionRepository: ReactionRepository,
    private val reactionValidator: ReactionValidator,
    private val applicationEventPublisher: ApplicationEventPublisher
) {
    @Transactional
    fun create(dto: ReactionDto): ReactionDto {
        reactionValidator.validate(dto)
        val reaction = reactionRepository.save(
            Reaction(
                target = ReactionTarget(dto.targetType, dto.targetId),
                type = dto.type,
                userId = dto.userId
            )
        )
        reaction.created(applicationEventPublisher)
        reactionRepository.save(reaction)
        return ReactionDto.from(reaction)
    }

    fun findById(id: Long): ReactionDto {
        val reaction = reactionRepository.findById(id).orElseThrow { ReactionNotFoundException() }
        return ReactionDto.from(reaction)
    }

    @Transactional
    fun delete(id: Long) {
        reactionRepository.deleteById(id)
    }
}
